lighthouse-core/lib/network-request.js

```diff
@@ -0,0 +1,260 @@
+    this.localizedFailDescription = data.errorText;
+
+    this._updateResponseReceivedTimeIfNecessary();
+  }
+
+  /**
+   * @param {LH.Crdp.Network.ResourceChangedPriorityEvent} data
+   */
+  onResourceChangedPriority(data) {
+    this.priority = () => data.newPriority;
+  }
+
+  /**
+   * @param {LH.Crdp.Network.RequestWillBeSentEvent} data
+   */
+  onRedirectResponse(data) {
+    if (!data.redirectResponse) throw new Error('Missing redirectResponse data');
+    this._onResponse(data.redirectResponse, data.timestamp, data.type);
+    this._resourceType = undefined;
+    this.finished = true;
+    this.endTime = data.timestamp;
+
+    this._updateResponseReceivedTimeIfNecessary();
+  }
+
+  /**
+   * @param {LH.Crdp.Network.Response} response
+   * @param {number} timestamp
+   * @param {LH.Crdp.Network.ResponseReceivedEvent['type']=} resourceType
+   */
+  _onResponse(response, timestamp, resourceType) {
+    this.url = response.url;
+    this._url = response.url;
+
+    this.connectionId = String(response.connectionId);
+    this.connectionReused = response.connectionReused;
+
+    if (response.protocol) this.protocol = response.protocol;
+
+    this._responseReceivedTime = timestamp;
+
+    this.transferSize = response.encodedDataLength;
+    if (typeof response.fromDiskCache === 'boolean') this._fromDiskCache = response.fromDiskCache;
+
+    this.statusCode = response.status;
+
+    this._timing = response.timing;
+    if (resourceType) this._resourceType = resourceTypes[resourceType];
+    this._mimeType = response.mimeType;
+    this._responseHeaders = NetworkRequest._headersDictToHeadersArray(response.headers);
+
+    this._fetchedViaServiceWorker = !!response.fromServiceWorker;
+
+    if (this._fromMemoryCache) this._timing = undefined;
+    if (this._timing) this._recomputeTimesWithResourceTiming(this._timing);
+  }
+
+  /**
+   * Resolve differences between conflicting timing signals. Based on the property setters in DevTools.
+   * @see https://github.com/ChromeDevTools/devtools-frontend/blob/56a99365197b85c24b732ac92b0ac70feed80179/front_end/sdk/NetworkRequest.js#L485-L502
+   * @param {LH.Crdp.Network.ResourceTiming} timing
+   */
+  _recomputeTimesWithResourceTiming(timing) {
+    // Take startTime and responseReceivedTime from timing data for better accuracy.
+    // Timing's requestTime is a baseline in seconds, rest of the numbers there are ticks in millis.
+    this.startTime = timing.requestTime;
+    const headersReceivedTime = timing.requestTime + timing.receiveHeadersEnd / 1000.0;
+    if (!this._responseReceivedTime || this._responseReceivedTime < 0) {
+      this._responseReceivedTime = headersReceivedTime;
+    }
+
+    this._responseReceivedTime = Math.min(this._responseReceivedTime, headersReceivedTime);
+    this._responseReceivedTime = Math.max(this._responseReceivedTime, this.startTime);
+    this.endTime = Math.max(this.endTime, this._responseReceivedTime);
+  }
+
+  _updateResponseReceivedTimeIfNecessary() {
```

Comment:
add a comment for what this is doing/when it's "necessary"?

Reply:
done